br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -247,59 +316,55 @@ func (manager *DuplicateManager) storeDuplicateData(
 	resp *import_sstpb.DuplicateDetectResponse,
 	decoder *kv.TableKVDecoder,
 	req *DuplicateRequest,
-) ([][]byte, error) {
-	opts := &pebble.WriteOptions{Sync: false}
+) (pendingIndexHandles, error) {
 	var err error
-	maxKeyLen := 0
+	var dataConflictInfos []errormanager.DataConflictInfo
+	indexHandles := makePendingIndexHandlesWithCapacity(len(resp.Pairs))
 	for _, kv := range resp.Pairs {
-		l := manager.keyAdapter.EncodedLen(kv.Key)
-		if l > maxKeyLen {
-			maxKeyLen = l
-		}
-	}
-	buf := make([]byte, maxKeyLen)
-	for i := 0; i < maxRetryTimes; i++ {
-		b := manager.db.NewBatch()
-		handles := make([][]byte, 0)
-		for _, kv := range resp.Pairs {
-			if req.indexInfo != nil {
-				h, err := decoder.DecodeHandleFromIndex(req.indexInfo, kv.Key, kv.Value)
-				if err != nil {
-					log.L().Error("decode handle error from index",
-						zap.Error(err), logutil.Key("key", kv.Key),
-						logutil.Key("value", kv.Value), zap.Uint64("commit-ts", kv.CommitTs))
-					continue
-				}
-				key := decoder.EncodeHandleKey(h)
-				handles = append(handles, key)
-			} else {
-				encodedKey := manager.keyAdapter.Encode(buf, kv.Key, 0, int64(kv.CommitTs))
-				b.Set(encodedKey, kv.Value, opts)
-			}
+		logger := log.With(
+			logutil.Key("key", kv.Key), logutil.Key("value", kv.Value),
+			zap.Uint64("commit-ts", kv.CommitTs))
+
+		var h tidbkv.Handle
+		if req.indexInfo != nil {
+			h, err = decoder.DecodeHandleFromIndex(req.indexInfo, kv.Key, kv.Value)
+		} else {
+			h, err = decoder.DecodeHandleFromTable(kv.Key)
```

Comment:
It seems that the object `handle` will hold a slice of `kv.Key`. I'm not sure it is a correct method to store `h` itself in other array.

Reply:
`kv.Key` is derived from the gRPC response `*import_sstpb.DuplicateDetectResponse`. In gogoproto, the `.Unmarshal` method will not reuse any buffer:

```go
func (m *KvPair) Reset()         { *m = KvPair{} }

func (m *KvPair) Unmarshal(dAtA []byte) error {
          ...
			m.Key = append(m.Key[:0], dAtA[iNdEx:postIndex]...)
          ...
}
```

so the `kv.Key` content won't be mutated outside of control.

Reply:
OK